### PR TITLE
[WOL] Implement wake on LAN command line tool

### PR DESCRIPTION
**Pull request by w1nda**
#### Why I did it
Implement wake on LAN command line tool to send magic packets to target interface with extra parameters like mac address, password, count and interval.

##### Work item tracking
- Microsoft ADO **(number only)**: 28600376

#### How I did it
Setup rust compile environment, write code and test to implement the tool.
Set raw socket capacity for wol binary when build image.

#### How to verify it
Unit test and manually test with packet capture.

wendachu@wenda-dev-vm:~/sonic-buildimage/src/sonic-nettools$ cargo test 
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.02s
     Running unittests src/main.rs (target/debug/deps/wol-a43f4ac2d14649cf)

running 9 tests
test wol::tests::test_build_magic_packet_without_password ... ok
test wol::tests::test_build_magic_packet ... ok
test wol::tests::test_is_ipv4_address_valid ... ok
test wol::tests::test_is_mac_string_valid ... ok
test wol::tests::test_parse_password ... ok
test wol::tests::test_parse_mac_addr ... ok
test wol::tests::test_parse_ipv4_addr ... ok
test wol::tests::test_parse_target_macs ... ok
test wol::tests::verify_args_parse ... ok

test result: ok. 9 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

File: sonic-slave-buster/Dockerfile.j2 [hunk 1]
```diff
@@ -644,7 +645,7 @@ RUN eatmydata apt-get install -y nodejs
...
...-dev:$arch libcap2-bin:$arch libpam0g-dev:$arch libaudit-dev:$arch libgtk-3-dev:$arch libkrb5-dev:$arch libsystemd-dev:$arch libwrap0-dev:$arch libkrb5-dev:$arch libboost1.67-dev:$arch libboost-dev:$arch libzmq5:$arch libzmq3-dev:$arch libdaemon-dev:$arch libjansson-dev:$arch libmnl-dev:$arch libsensors5:$arch libsensors4-dev:$arch libperl-dev:$arch libmariadb-dev:$arch libmariadb-dev-compat:$arch libpci-dev:$arch libjson-c3:$arch libjson-c-dev:$arch libreadline-dev:$arch librtr-dev:$arch librrd-dev:$arch libnetfilter-conntrack-dev:$arch libnetfilter-conntrack3:$arch libnfnetlink-dev:$arch libnftnl-dev:$arch libldap2-dev:$arch libbind-export-dev:$arch check:$arch libboost-atomic-dev:$arch libboost-test-dev:$arch libglib2.0-dev:$arch qt5-default:$arch libexplain-dev:$arch libc-ares-dev:$arch libip4tc0:$arch libip6tc0:$arch libiptc0:$arch libxtables12:$arch iptables-dev:$arch libatm1-dev:$arch libdb-dev:$arch pkg-config:$arch libnghttp2-14:$arch librtmp1:$arch libssh2-1:$arch libcjson1:$arch libcjson-dev:$arch libcurl4-openssl-dev:$arch libboost-thread1.67-dev:$arch libboost-thread-dev:$arch libboost-system1.67-dev:$arch libboost-system-dev:$arch libgtest-dev:$arch libgmock-dev:$arch
```
Comment (liushilongbuaa) [inline sonic-slave-buster/Dockerfile.j2:648]: Don't change jessie, buster, bullseye if you don't use it.

File: slave.mk [hunk 2]
```diff
@@ -362,6 +362,15 @@ CROSS_COMPILE_FLAGS := CGO_ENABLED=1 GOOS=linux GOARCH=$(GOARCH) CROSS_COMPILE=$
 
 endif
 
+ifeq ($(CROSS_BUILD_ENVIRON),y)
```
Comment (liushilongbuaa) [inline slave.mk:365]: What if not cross build env?

File: sonic-slave-bookworm/Dockerfile.j2 [hunk 3]
```diff
@@ -736,3 +738,15 @@ RUN eatmydata apt install -y python-is-python3
...
 RUN curl -fsSL -o /usr/local/bin/bazel ${bazelisk_url} && chmod 755 /usr/local/bin/bazel
 {% endif -%}
+
+# Install Rust
+ARG RUST_ROOT=/usr/.cargo
+RUN RUSTUP_HOME=$RUST_ROOT CARGO_HOME=$RUST_ROOT bash -c 'curl --proto "=https" -sSf https://sh.rustup.rs | sh -s -- --default-toolchain 1.79.0 -y'
+{% if CROSS_BUILD_ENVIRON == "y" and CONFIGURED_ARCH == "armhf" %}
+RUN mkdir -p /.cargo && $RUST_ROOT/bin/rustup target add armv7-unknown-linux-gnueabihf && echo "[target.armv7-unknown-linux-gnueabihf]\nlinker = \"arm-linux-gnueabihf-gcc\"" >> /.cargo/config.toml
+{% endif -%}
+{% if CROSS_BUILD_ENVIRON == "y" and CONFIGURED_ARCH == "arm64" %}
```
Comment (liushilongbuaa) [inline sonic-slave-bookworm/Dockerfile.j2:748]: CROSS_BUILD_ENVIRON means that build armhf on amd64 or build arm64 on amd64.
Currently we build armhf on armhf and build arm64 on amr64.

File: sonic-slave-bookworm/Dockerfile.j2 [hunk 4]
```diff
@@ -736,3 +738,15 @@ RUN eatmydata apt install -y python-is-python3
 ARG bazelisk_url=https://github.com/bazelbuild/bazelisk/releases/latest/download/bazelisk-linux-{{ CONFIGURED_ARCH }}
 RUN curl -fsSL -o /usr/local/bin/bazel ${bazelisk_url} && chmod 755 /usr/local/bin/bazel
 {% endif -%}
+
+# Install Rust
+ARG RUST_ROOT=/usr/.cargo
+RUN RUSTUP_HOME=$RUST_ROOT CARGO_HOME=$RUST_ROOT bash -c 'curl --proto "=https" -sSf https://sh.rustup.rs | sh -s -- --default-toolchain 1.79.0 -y'
+{% if CROSS_BUILD_ENVIRON == "y" and CONFIGURED_ARCH == "armhf" %}
```
Comment (liushilongbuaa) [inline sonic-slave-bookworm/Dockerfile.j2:745]: What about CROSS_BUILD_ENVIRON = n?

Comment (w1nda) [reply sonic-slave-bookworm/Dockerfile.j2:745]: If CROSS_BUILD_ENVIRON = y, we will install toolchains for target architecture,
else, we won't.

File: sonic-slave-buster/Dockerfile.j2 (hunk 1 above)
Comment (w1nda) [reply sonic-slave-buster/Dockerfile.j2:648]: My concern is if I don't install rust on jessie, buster and bullseye, everybody will fail to build due to lack of rust toolchain.

So, I installed the rust on those docker images.

Comment (w1nda) [reply sonic-slave-buster/Dockerfile.j2:648]: Offline synced with @liushilongbuaa, we will modify bookworm on master branch, modify others when backport.

File: src/sonic-nettools/wol/src/wol.rs [hunk 5]
```diff
@@ -0,0 +1,687 @@
...
+        assert_eq!(magic_packet.len(), 116);
+        assert_eq!(&magic_packet[0..6], &dst_mac);
+        assert_eq!(&magic_packet[6..12], &src_mac);
+        assert_eq!(&magic_packet[12..14], &[0x08, 0x42]);
+        assert_eq!(&magic_packet[14..20], &[0xff; 6]);
+        assert_eq!(&magic_packet[20..116], target_mac.repeat(16));
+    }
+
+    #[test]
+    fn verify_args_parse() {
```
Comment (lizhijianrd) [inline src/sonic-nettools/wol/src/wol.rs:555]: It's better to split this testcase to several testcases, each one for a single argument or scenario.

**Review by lizhijianrd — APPROVED** [review]

**Comment by lizhijianrd** [comment]
Can you help answer below questions in PR description:
1. Build/Compile environment is cleaned up, which means SONiC image doesn't contain the build env.
2. Paste unittest result to `How to verify it` section.
3. It the manually functionality test passed on both x86 and armhf?

**Comment by w1nda** [comment]
> Can you help answer below questions in PR description:
> 
> 1. Build/Compile environment is cleaned up, which means SONiC image doesn't contain the build env.
> 2. Paste unittest result to `How to verify it` section.
> 3. It the manually functionality test passed on both x86 and armhf?

1. Yes, the SONiC image doesn't contain the build env.
2. Updated.
3. Yes, the unit tests were ran in buildimage pipeline, and it passed on all hardware sku.

File: slave.mk (hunk 2 above)
Comment (liushilongbuaa) [inline slave.mk:365]: I'm not sure if you understand CROSS_BUILD_ENVIRON correctly.
Do you try to build arm SONiC image on amd64 machine?

Comment (w1nda) [reply slave.mk:365]: No, I'm trying install toolchain if user want to do cross build.

**Review by liushilongbuaa — APPROVED** [review]

**Comment by w1nda** [comment]
Hi @yxieca, cloud you help merge this PR?

**Comment by w1nda** [comment]
Cherry-pick to 202405 successfully, however, the labels are not updated,
I will update it manually.

**Pull request merged**